How To - Use Redis for memory (#40539)
dotnet/docs: docs/ai/how-to/use-redis-for-memory.md

---
title: "Use Redis for Memory Storage with the Semantic Kernel SDK for .NET"

> description: "Learn how to use a Redis database with the RediSearch module for memory storage and recall in Semantic Kernel SDK for .NET."

Author: I can add an explanation about memories to the introduction. I don't think it can be expressed concisely enough for the description. But I could reword this like:

> Learn how to use a Redis database with the RediSearch module to give your AI memories in Semantic Kernel SDK for .NET.

---
title: "Use Redis for Memory Storage with the Semantic Kernel SDK"
description: "Learn how to use a Redis database with the RediSearch module to give your AI memories in Semantic Kernel SDK for .NET."
author: haywoodsloan
ms.topic: how-to
ms.date: 04/17/2024

#customer intent: As a .NET developer, I want to use Redis for memory storage with the Semantic Kernel SDK so that I can store and recall memories in my application.

---

# Use Redis for memory storage with the Semantic Kernel SDK

This article demonstrates how to integrate a Redis database with the RediSearch module into the [Semantic Kernel SDK](/semantic-kernel/overview) and use it for memory storage and retrieval.

> [Memories](/semantic-kernel/memories/) represent text information that has be stored alongside a precomputed embedding vector for the whole text. When an LLM is prompted to recall a memory it uses these precomputed embeddings to efficiently evaluate if a memory is relevant to the prompt. Once the LLM finds a matching memory it uses the memory's text information as context for the next steps in the prompt completion.

Reviewer: ```suggestion
[Memories](/semantic-kernel/memories/) represent text information that has been stored alongside a precomputed embedding vector for the whole text. When an LLM is prompted to recall a memory, it uses these precomputed embeddings to efficiently evaluate if a memory is relevant to the prompt. Once the LLM finds a matching memory, it uses the memory's text information as context for the next steps in the prompt completion.
```

> Adding memory storage to the Semantic Kernel SDK provides a broader context for your requests, and enables you to store data like a traditional database but query it using natural language.

Reviewer: ```suggestion
When you add memory storage to the Semantic Kernel SDK, it provides a broader context for your requests. It also enables you to store data like a traditional database but query it using natural language.
```

## Prerequisites

* An Azure account with an active subscription. [Create an account for free](https://azure.microsoft.com/free/?WT.mc_id=A261C142F).
* [.NET SDK](https://dotnet.microsoft.com/download/visual-studio-sdks)
* [`Microsoft.SemanticKernel` NuGet package](https://www.nuget.org/packages/Microsoft.SemanticKernel)
* [`Microsoft.SemanticKernel.Connectors.Redis` NuGet package](https://www.nuget.org/packages/Microsoft.SemanticKernel.Connectors.Redis)
* [`Microsoft.SemanticKernel.Plugins.Memory` NuGet package](https://www.nuget.org/packages/Microsoft.SemanticKernel.Plugins.Memory)
* [`StackExchange.Redis` NuGet package](https://www.nuget.org/packages/StackExchange.Redis)

> * A Redis database with the RediSearch module, [deployed and accessible to your .NET application](/azure/azure-cache-for-redis/quickstart-create-redis-enterprise)

Reviewer 2: ```suggestion
* A Redis database that has the RediSearch module, [deployed and accessible to your .NET application](/azure/azure-cache-for-redis/quickstart-create-redis-enterprise)
```
Additional change: Currently, this sentence indicates that the Redis database is deployed and accessible. Is that what we mean? If we mean that the RediSearch module is deployed and accessible, please replace the comma after "module" with "that's" ("module that's deployed").

Author: Is the Redis database that's deployed and accessible it just also needs the RediSearch module enabled.

## Implement memory storage using a Redis database

> Before integrating your Redis database to the Semantic Kernel SDK, ensure you have the RediSearch module enabled. For _Azure Cache for Redis_ see [Use Redis modules with Azure Cache for Redis](/azure/azure-cache-for-redis/cache-redis-modules#adding-modules-to-your-cache).

Reviewer: ```suggestion
Before integrating your Redis database to the Semantic Kernel SDK, ensure you have the RediSearch module enabled. For module information for _Azure Cache for Redis_, see [Use Redis modules with Azure Cache for Redis](/azure/azure-cache-for-redis/cache-redis-modules#adding-modules-to-your-cache).
```

> Before integrating your Redis database to the Semantic Kernel SDK, ensure you have the RediSearch module enabled. For module information for _Azure Cache for Redis_, see [Use Redis modules with Azure Cache for Redis](/azure/azure-cache-for-redis/cache-redis-modules#adding-modules-to-your-cache).

Reviewer 2: ```suggestion
Before you integrate your Redis database to the Semantic Kernel SDK, make sure that you have the RediSearch module enabled. For module information for _Azure Cache for Redis_, see [Use Redis modules with Azure Cache for Redis](/azure/azure-cache-for-redis/cache-redis-modules#adding-modules-to-your-cache).
```

1. Initialize a connection to your Redis database. For example:

    :::code language="csharp" source="./snippets/semantic-kernel/MemoryExamples.cs" id="initRedis":::

> 2. Build the `Kernel` with a `ITextEmbeddingGenerationService` included. For example:

Reviewer 2: ```suggestion
2. Build the `Kernel` by including `ITextEmbeddingGenerationService`. For example:
```

    :::code language="csharp" source="./snippets/semantic-kernel/MemoryExamples.cs" id="initKernel":::

> 3. Wrap the Redis database in a `RedisMemoryStore`, then initialize a `SemanticTextMemory` using the memory store and embedding generation service. For example:

Reviewer: ```suggestion
3. Wrap the Redis database in a `RedisMemoryStore`, then initialize a `SemanticTextMemory` object using the memory store and embedding generation service. For example:
```

> 3. Wrap the Redis database in a `RedisMemoryStore`, then initialize a `SemanticTextMemory` object using the memory store and embedding generation service. For example:

Reviewer 2: ```suggestion
3. Wrap the Redis database in a `RedisMemoryStore` instance, and then initialize a `SemanticTextMemory` object by using the memory store and embedding generation service. For example:
```
We need a label for `RedisMemoryStore`. Is "instance" accurate?

Author: yeah, instance works here

    :::code language="csharp" source="./snippets/semantic-kernel/MemoryExamples.cs" id="initMemory":::

> 4. Add the semantic text memory to the `Kernel` using the `TextMemoryPlugin`. For example:

Reviewer 2: ```suggestion
4. Add the semantic text memory to the `Kernel` by using the `TextMemoryPlugin` class. For example:
```
Ditto. Is "class" accurate?

Alt: "...by using `TextMemoryPlugin`. (Probably a better choice if this element is correctly labeled as a plug-in.)

Author: In terms of .Net it's a class

    :::code language="csharp" source="./snippets/semantic-kernel/MemoryExamples.cs" id="addMemory":::

> 5. Use the `Kernel` and plugin to save, retrieve, and recall memories. For example:

Reviewer 2: ```suggestion
5. Use the `Kernel` and plug-in to save, retrieve, and recall memories. For example:
```

    :::code language="csharp" source="./snippets/semantic-kernel/MemoryExamples.cs" id="useMemory":::

> 6. Use memory recall as part of a prompt with the [prompt template syntax `{{...}}`](/semantic-kernel/prompts/prompt-template-syntax). For example:

Reviewer 2: ```suggestion
6. Use memory recall as part of a prompt that has the [prompt template syntax `{{...}}`](/semantic-kernel/prompts/prompt-template-syntax). For example:
```
Alt (depending on meaning): "as part of a prompt by using..."
